tket/src/Circuit/ToffoliBox.cpp

```diff
@@ -0,0 +1,320 @@
+  // At jth column j, for any row R whose jth bit doesn't match the jth bit of
+  // its row index, swap it with row R' where R and R' only differ at the jth
+  // bit. All the swaps can be done with one multiplexor targeting q[j].
+
+  // In the implementation, we use a map from rows to rows indices to represent
+  // the matrix defined in the algorithm.
+
+  Circuit circ(n_qubits);
+  // special case
+  if (n_qubits == 1) {
+    if (perm.begin()->first != perm.begin()->second) {
+      circ.add_op<unsigned>(OpType::X, {0});
+    }
+    return circ;
+  }
+  // track relative phases caused by using pi rotations
+  std::vector<Complex> phases(1 << n_qubits, 1.);
+  // step 1
+  for (unsigned col_idx = 0; col_idx < n_qubits - 1; col_idx++) {
+    ctrl_op_map_t op_map;
+    for (unsigned prefix = 0; prefix < (unsigned)(1 << col_idx); prefix++) {
+      std::vector<bool> prefix_bin;
+      if (col_idx != 0) {
+        prefix_bin = dec_to_bin(prefix, col_idx);
+      }
+      std::vector<std::vector<bool>> swap_pairs =
+          rearrange_along_col(prefix_bin, n_qubits, col_idx, perm);
+      for (auto pair : swap_pairs) {
+        swap_rows(pair, col_idx, perm, op_map, phases, zflip_op);
+      }
+    }
+    if (!op_map.empty()) {
+      std::vector<unsigned> multplx_args =
+          get_multiplexor_args(n_qubits, col_idx);
+      circ.add_box(MultiplexedRotationBox(op_map), multplx_args);
+    }
+  }
+  // step 2
+  for (unsigned col_idx = n_qubits; col_idx-- > 0;) {
+    ctrl_op_map_t op_map;
+    for (unsigned row = 0; row < (unsigned)(1 << (n_qubits - 1)); row++) {
+      std::vector<bool> pair = dec_to_bin(row, n_qubits - 1);
+      pair.insert(pair.begin() + col_idx, 0);
+      if (perm[pair][col_idx] != 0) {
+        pair.erase(pair.begin() + col_idx);
+        swap_rows(pair, col_idx, perm, op_map, phases, zflip_op);
+      }
+    }
+    if (!op_map.empty()) {
+      std::vector<unsigned> multplx_args =
+          get_multiplexor_args(n_qubits, col_idx);
+      circ.add_box(MultiplexedRotationBox(op_map), multplx_args);
+    }
+  }
+  // correct the phases with a diagonal operator
+  Eigen::VectorXcd corrections(1 << n_qubits);
+  for (unsigned i = 0; i < phases.size(); i++) {
+    corrections[i] = 1. / phases[i];
+  }
+  DiagonalBox diag(corrections);
+  circ.add_box(diag, circ.all_qubits());
+  return circ;
+}
+
+void ToffoliBox::generate_circuit() const {
+  unsigned n_qubits = (unsigned)permutation_.begin()->first.size();
+  state_perm_t perm(permutation_);
+  // fill the permutation with identities
+  if ((unsigned)perm.size() != (unsigned)(1 << n_qubits)) {
+    for (unsigned i = 0; i < (unsigned)(1 << n_qubits); i++) {
+      std::vector<bool> state = dec_to_bin(i, n_qubits);
+      if (perm.find(state) == perm.end()) {
+        perm.insert({state, state});
+      }
+    }
+  }
+  Circuit circ = permute(perm, n_qubits, get_op_ptr(rotation_axis_, 1));
+  circ_ = std::make_shared<Circuit>(circ);
```

Comment:
Minor comment, any reason to not do:
`circ_ = std::make_shared<Circuit>(permute(perm(n_qubits, get_op_ptr(rotation_axis_, 1));`?